src/wasm-lib/kcl/src/execution/memory.rs

```diff
@@ -1,74 +1,405 @@
+//! Representation of KCL memory.
+//!
+//! Stores `KclValue`s by name using dynamic scoping. Memory does not support addresses or references,
+//! so all values must be self-contained. Memory is essentially a map from `String`s to `KclValue`s.
+//! `KclValue`s are entirely opaque to this module. Memory is global and there should be only
+//! one per execution. It has no explicit support for caching between executions.
+//!
+//! Memory is mostly immutable (since KCL does not support mutation or reassignment). However, tags
+//! may change as code is executed and that mutates memory. Therefore,
+//! ProgramMemory supports mutability and does not rely on KCL's (mostly) immutable nature.
+//!
+//! ProgramMemory is observably monotonic, i.e., it only grows and even when we pop a stack frame the
+//! frame, it is retained. We remove some values which we know cannot be referenced, but we
+//! should in the future do better garbage collection (of values and envs).
+//!
+//! ## Concepts
+//!
+//! There are three main moving parts for ProgramMemory: environments, snapshots, and the call stack. I'll
+//! cover environments (and the call stack) first as if snapshots didn't exist, then describe snapshots.
+//!
+//! An environment is a set of bindings (i.e., a map from names to values). Environments handle
+//! both scoping and context switching. A new lexical scope means a new environment. Nesting of scopes
+//! means that environments form a tree, which is represented by parent pointers in the environments.
```

Comment:
A tree, or a list? Can there be two different environments that share a parent both instantiated at one time? Or if there's only ever 1 such environment then it's more like a stack right?

Reply:
In this example, we make two closures, `f` and `g`. They both must refer to the outer scope where `x` is defined. `f`'s scopes are a list, and `g`'s scopes are a list, but all together, it's a tree.

```kcl
x = 1

fn inner() {
  fn f() {}
  fn g() {}
  return [f, g]
}
```

```
     root
      |
    inner
     /  \
   f     g
```